File: src/learning_rosbag_cpp/src/bag_recoder.cpp
/**
 * 需求：录制 turtle_teleop_key 节点发布的速度指令
 * 步骤：
 *      1、包含头文件
 *      2、初始化 ROS2 客户端
 *      3、自定义节点类
 *          3-1、创建写出对象指针和订阅对象指针
 *          3-2、打开目标文件
 *          3-3、写出订阅到的数据
 *      4、调用 spin 函数 并传入节点对象指针
 *      5、释放资源
 */

#include "geometry_msgs/msg/twist.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rosbag2_cpp/writer.hpp"

using geometry_msgs::msg::Twist;
using std::placeholders::_1;

class BagRecoder : public rclcpp::Node {
public:
    BagRecoder() : Node("bag_recoder_cpp") {
        RCLCPP_INFO(this->get_logger(), "数据录制方创建成功!");
        writer_ = std::make_unique<rosbag2_cpp::Writer>();
        writer_->open("my_bag_cpp");
        subscriber_ = this->create_subscription<Twist>(
            "/turtle1/cmd_vel", 10,
            std::bind(&BagRecoder::write_callback ,this, _1)
        );
    }

private:
    void write_callback(const std::shared_ptr<rclcpp::SerializedMessage> msg) {
        RCLCPP_INFO(this->get_logger(), "数据写入中...");
        /*
            void write(
                std::shared_ptr<rclcpp::SerializedMessage> message,
                const std::string &topic_name,
                const std::string &type_name,
                const rclcpp::Time &time
            )
        */
        writer_->write(msg, "/turtle1/cmd_vel", "geometry_msgs/msg/Twist", this->now());
    }

    std::unique_ptr<rosbag2_cpp::Writer> writer_;
    rclcpp::Subscription<Twist>::SharedPtr subscriber_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<BagRecoder>());
    rclcpp::shutdown();

    return 0;
}

File: src/learning_rosbag_py/learning_rosbag_py/bag_recoder.py
"""  
  需求：录制 turtle_teleop_key 节点发布的速度指令。
  步骤：
    1.导包；
    2.初始化 ROS 客户端；
    3.定义节点类；
      3-1.创建写出对象；
      3-2.设置写出的目标文件、话题等参数；
      3-3.写出消息。
    4.调用 spin 函数，并传入对象；
    5.释放资源。

"""
# 1.导包；
import rclpy
from rclpy.node import Node
from rclpy.serialization import serialize_message
from geometry_msgs.msg import Twist
import rosbag2_py
# 3.定义节点类；
class SimpleBagRecorder(Node):
    def __init__(self):
        super().__init__('simple_bag_recorder_py')
        # 3-1.创建写出对象；
        self.writer = rosbag2_py.SequentialWriter()
        # 3-2.设置写出的目标文件、话题等参数；
        storage_options = rosbag2_py._storage.StorageOptions(
            uri='my_bag_py',
            storage_id='sqlite3')
        converter_options = rosbag2_py._storage.ConverterOptions('', '')
        self.writer.open(storage_options, converter_options)

        topic_info = rosbag2_py._storage.TopicMetadata(
            name='/turtle1/cmd_vel',
            type='geometry_msgs/msg/Twist',
            serialization_format='cdr')
        self.writer.create_topic(topic_info)

        self.subscription = self.create_subscription(
            Twist,
            '/turtle1/cmd_vel',
            self.topic_callback,
            10)
        self.subscription

    def topic_callback(self, msg):
        # 3-3.写出消息。
        self.writer.write(
            '/turtle1/cmd_vel',
            serialize_message(msg),
            self.get_clock().now().nanoseconds)


def main(args=None):
    # 2.初始化 ROS 客户端；
    rclpy.init(args=args)
    # 4.调用 spin 函数，并传入对象；
    sbr = SimpleBagRecorder()
    rclpy.spin(sbr)
    # 5.释放资源。
    rclpy.shutdown()


if __name__ == '__main__':
    main()